work in progress on v1

wadcom opened the pull request:
This is a work in progress, there are still some features to add.

The pull request is based off master, so it contains everything committed to the project so far, for reviewing convenience.

wadcom committed:
Record initial requirements in a TODO/Backlog

wadcom committed:
Reimport all prework to put it into a PR

Reimported all files as submitted the first time. This is to make it
possible to create a PR on GitHub for more convenient review.

This squashes the history, but that's of not much value at this point.

wadcom committed:
Fix: MailQueue(fresh=True) was failing if database file didn't exist

wadcom committed:
Move tests from frontend.py into test_frontend.py

wadcom committed:
Move tests from mailqueue.py into test_mailqueue.py

wadcom committed:
Update file pattern for unittests discovery

wadcom committed:
Add a draft of update frontend design

This lays groundwork for meeting the new requirements: delivery to
multiple SMTP servers based on MX records, querying delivery status
with a retention policy, retries.

wadcom committed:
Add missing diagrams

wadcom committed:
Add a sketch of the sender process design

wadcom committed:
Reimplement end-to-end test in Python using unittest

With the new requirements Shell-based test won't be easy to
read/maintain, so it makes sense to convert it to Python right away.

We could use something like Behave, but unittest will do for now (and
spares us from introducing another dependency).

wadcom committed:
Hide MiniMailGun endpoint URL inside Client

Client objects will be used in multiple testcases, so passing the
endpoint URL as a constructor parameter would cause duplication.

wadcom committed:
Get domain name from environment and use it as a part of log file name

This will be useful when running multiple instances of SMTP stubs
during end-to-end tests.

wadcom committed:
MailQueue.put() now accepts a mailqueue.Envelope

wadcom committed:
MailQueue now returns Envelope objects

mailqueue.Item is removed.

wadcom committed:
Remove all SMTP stubs' logs

wadcom committed:
SendHandler now properly generates envelopes with destination_domain

Recipients should now be passed as a list of strings (email addresses).
SendHandler now groups recipients by destination domain and creates
corresponding number of envelopes.

wadcom committed:
Sender now uses destination_domain from envelope to find MXs

This is still a prototype code, just to see how it falls out.
Single-threaded, one-message-at-a-time, no error handling. Will clean
up once the structure emerges.

wadcom committed:
Add an end-to-end test for delivery to different domains

wadcom committed:
Cosmetics: restructure the loop to avoid indentation

wadcom committed:
Factor out (still dirty) SMTP client logic into SMTPClient class

wadcom committed:
Add support for "next scheduled attempt" for an envelope

MailQueue now has method schedule_retry_in(). The database now selects
envelopes which are already up for next delivery attempt.

wadcom committed:
Make stdout/stderr unbuffered

Without this both streams are buffered when run by docker-compose,
which makes diagnostics difficult.

wadcom committed:
A first cut of factoring out delivery agent logic

Failures happening during SMTP session are not handled (only MX lookup
failures are).

wadcom committed:
Stand up 2 SMTP server stubs

wadcom committed:
Add an end-to-end test for partial delivery

wadcom committed:
Add a TODO note re optimizing the # of outbound SMTP connections

wadcom committed:
Add logging

wadcom committed:
Implement simplistic MX records lookup

Only first MX is used, no looping through MX records occur. It is now
possible to override MX resolution via DNS with STATIC_MX_CONFIG
environment variable. Default SMTP port can also be overridden with
SMTP_PORT.

wadcom committed:
Fix: MailQueue.schedule_retry_in() based next retry on previous attempt

Another retry was scheduled with specified offset from the _scheduled_
time of previous retry, not the actual one.

In order to test these kinds of scenarios I had to pull
timestamp-related logic from the database (e.g. I no longer rely on
DB's default value for next_attempt_at column). It's more brittle in
theory, but the DB is currently embedded in the same process (thus
sharing the clock), and this kind of setup allows to extensively test
various timing scenarios, soI think it's worth it in the current
context.

wadcom committed:
Factor out Handler._handle_post_with()

SendHandler.run() now returns the response dictionary to match the new
contract.

wadcom committed:
Introduce mailqueue.Manager() to synchronize access to MailQueue

wadcom committed:
Make use of mailqueue.Manager

Removed now obsolete MailQueueAppender

wadcom committed:
Make _make_envelopes and _make_message methods static

wadcom committed:
Remove now unneded 'self' parameter from _make_envelopes

wadcom committed:
Factor out make_valid_email and make_valid_envelope into testhelpers

wadcom committed:
Add support for 'status' and 'submission_id' fields in mail queue

Added MailQueue.get_status() (by submission_id) and proxied this
message in Manager. Also added NOT NULL contraints to DB columns.

wadcom committed:
Add /status endpoint

The /send endpoint now returns submission id.

wadcom committed:
Handle multiple envelopes in a single submission in get_status()

Also, MailQueue.put() now returns the id of newly added envelope.

wadcom committed:
Added SENT status

wadcom committed:
Aggregate envelopes' statuses when getting submission status

wadcom committed:
Sent envelopes are no longer deleted from the database

Their status is changed, so that /status endpoint can return the status
appropriately.

wadcom committed:
Handle SMTP session temporary failures

Also, retry interval is now settable in DeliveryAgent. Sender uses
RETRY_INTERVAL environment variable to override the default.

wadcom committed:
Simulate temporary failures

The stub now simulates a temporary failure is the sender address takes
special form, for use in end-to-end tests.

wadcom committed:
Add end-to-end test for delivery retries

wadcom committed:
Update end-to-end test structure diagram

wadcom committed:
Update HTTP endpoints documentation

wadcom committed:
Update system design documentation

This reflects a simplier design than the initial one.

wadcom committed:
Update high-level service description

wadcom committed:
Update TODO

wadcom committed:
Remove unused entities diagram

wadcom committed:
Add 'delivery_attempts' column

wadcom committed:
Add MailQueue.mark_as_undeliverable(), keep track of # of retries

wadcom committed:
Add 'delivery_attempts' to Envelope

wadcom committed:
Limit the number of delivery attempts

Mark the envelope as undeliverable after reaching the limit.

wadcom committed:
Add an end-to-end test for limiting the number of retries

wadcom committed:
Remove completed to-do item

obukhov-sergey commented on e2e-tests/e2e-test.py:
```diff
@@ -0,0 +1,141 @@
+
+
+if __name__ == '__main__':
+    unittest.main()
```
nit pick - a new line is missing

wadcom replied:
Fixed, thanks!

obukhov-sergey commented on sender.py:
```diff
@@ -0,0 +1,150 @@
+        self._smtp_client = smtp_client
+
+    def deliver_single_envelope(self):
+        envelope = self._mailqueue.get()
```
If you're still looking for last minute requests I'd ask for scaling senders.

wadcom replied:
I'm currently working on making the sender multi-threaded. This will help with scale, but we can go further be running multiple `sender` processes (this will require sharding).

Do you want me to work on that before completing other items in the TODO?

obukhov-sergey replied:
If multi-threaded sender is close to completion I'd finish it first. Curious how it's gonna look like. And I'd do the multiple sender processes next.

wadcom replied:
@obukhov-sergey, implemented both multi-threading and sharded senders, please take a look.

obukhov-sergey commented on frontend.py:
```diff
@@ -0,0 +1,127 @@
+        else:
+            self.send_error(404, 'Not found')
+
+    def _handle_post_with(self, handler_class):
```
It's a pretty smart way to handle requests. It's pretty clean and there is no code duplication. But I'd still rather see ``response_data = SendHandler.run(body)`` than ``response_data = handler_class(mq_manager).run(body)``. It can result in more code and even in code duplication but I'd have to think less :)

wadcom replied:
Hmm, I'd be happy to improve readability (although the way it's done now doesn't strike me as poorly readable), and to do that I'd like to understand what aspect of the code seems stumbling to you. Is it passing the class object around? Or a chained call to the class constructor and then `.run()`?

Will something like this look better:

```
    def do_POST(self):
        if self.path == '/send':
            handler = SendHandler(mq_manager)
        elif self.path == '/status':
            handler = StatusHandler(mq_manager)
        else:
            self.send_error(404, 'Not found')

       self._handle_post_with(handler)
```
?

obukhov-sergey replied:
Passing the class object. The chained call is fine. Yep, I like the suggested variant better.

wadcom replied:
Done!

obukhov-sergey reviewed (commented):
Great job! Enjoyed the README and TODO files. Loved the hand-written diagrams: simple and effective. The code has some smart patterns that make it less readable. But it's not over-complicated. It's also well tested, clean. Found no obvious flaws.

wadcom committed:
Add missing trailing newline

wadcom committed:
Automatically detect and proxy all MailQueue's methods in Manager

This is a bit more complex, but spares us from manually duplicating all
MailQueue's methods in Manager.

anton-efimenko commented on mailqueue.py:
```diff
@@ -0,0 +1,181 @@
+            assert method not in self.__class__.__dict__, \
+                "name clash: method '{}' already exists in {}".format(method,
+                                                                      self.__class__.__name__)
+            self.__dict__[method] = self.ProxiedMethod(method, self._requests, self._responses)
```
Could you explain what it is for?

wadcom replied:
Manager is a drop-in replacement for MailQueue. So we need it to have all the same methods as MailQueue.

One option is to manually implement each of them, but this means every change in MailQueue will lead to the same kind of change in Manager.

The other option (implemented here) is that Manager inspects the MailQueue class and dynamically creates all public methods (which proxy requests to MailQueue). This way Manager's methods are always up-to-date with MailQueue's and we don't have to manually replicate them each time there is a change in MailQueue.

anton-efimenko replied:
As far as I understood the Manager implements a synchronization mechanism for the MailQueue. I understand how it works but wouldn't things like `self._responses.put(getattr(mq, method)(*args))` be painful to debug when they throw an error?

wadcom replied:
So, return values which indicate errors will be handled by callers of Manager.some_method(), so it shouldn't be an issue.

If a MailQueue method raises an exception, here's how it will look:

```
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/app/mailqueue.py", line 174, in _main_loop
    self._responses.put(getattr(mq, method)(*args))
  File "/app/mailqueue.py", line 31, in raise_exception
    raise NotImplementedError
NotImplementedError
```
It does indicate MailQueue's method which raised the exception, so not all is lost (although we wouldn't know who issued the original call and with which arguments).

I don't see how we can improve diagnostics (without further complicating the mechanism, like passing a stack trace along with method arguments to the Manager thread) in a design where Manager runs in a separate thread.

SQLite has a requirement that a database connection should be only used by the thread which created it, thus warranting such design. There is a flag to relax this restriction, but I haven't looked closely into implications of its use.

The code for intra-thread communication with MailQueue is pretty simplistic and it provides *some* diagnostics which tend to be enough in practice (I usually write a failing end-to-end test before writing the code, so I have a chance to observe various ways how execution might fail as I gradually proceed with implementation).

So I suggest to leave it as is for now and get back to it later. Unless, of course, I'm missing something. What would be your suggestion?

anton-efimenko replied:
That was precisely my point. Although it looks succinct in the reality you'll have a hard time debugging it. Especially if that needs to be done quickly to mitigate production downtime.

Regarding the solution, if we decided to stick with SQLite (which is clearly limiting us) I would opt in for the following options:
* Use `check_same_thread=False` flag you mentioned. The downside is that you'll have to handle concurrency in the code.
* ~Implement task-agnostic client to the database that manages the serializations of the queries.~  It already exists: https://github.com/palantir/sqlite3worker
* I haven't looked deep enough but it seems the SQLite can be compiled with `SQLITE_THREADSAFE=1` option. Not sure if the python client supports it.

All these options have their own drawbacks and I agree that we can leave it as is. The important thing is that we strive for simplicity because it could be incredibly hard to debug another person's code under time pressure.

wadcom replied:
I agree, if we stick with SQLite (which might be ok for the initial launch, but will probably have to be replaced with something else down the road), then `sqlite3worker` seems like the best way to go — good find!

anton-efimenko reviewed (commented):
Good job! However I would caution you against over-engineering.

wadcom committed:
Mark envelopes already taken for processing

This implementation doesn't use transactions assuming callers will
synchronize (e.g. by using mailqueue.Manager).

Also, if the process crashes, the envelope will remain marked as
"being processed". But this simplistic implementation will work as a
first cut.

wadcom committed:
Implement multiple delivery threads

wadcom committed:
Simulate slow response from SMTP server

wadcom committed:
Add an end-to-end test for concurrent delivery

wadcom committed:
_handle_post_with() now accepts handler object, not a class

wadcom committed:
MailQueue.get() will now retrieve envelopes of correct shard

Sharding configuration is passed via shard= and shards= keywords to the
constructor.

wadcom committed:
Implement strict checks of the number of messages delivered

To achieve this we have to give opportunistic wait: we now always wait
for specified timeout before checking the logs.

wadcom committed:
Add support for sharing reads from mailqueue

Sharding configuration is passed via environment variable SHARD, e.g.
SHARD=2/3 (1-based, not 0-based).

wadcom committed:
Start 2 sender shards for end-to-end tests

wadcom committed:
Mention sharded sender configuration

wadcom committed:
Reduce the retry interval to 3 seconds to speed up tests

wadcom committed:
Add 'client_id' to the database, make get_status() honor it

Now get_status() will only return status for this client's envelopes.

wadcom committed:
Add client_id to the stock envelope

wadcom committed:
Enforce client authentication

The list of valid client ids is taken from /conf/clients.

wadcom committed:
Add end-to-end tests for authentication and authorization

wadcom committed:
Remove completed items

wadcom committed:
Remove redundant 'sender' column from the database

The message sender is already stored in the 'From' header of the
message.

wadcom committed:
Add MailQueue.remove_inactive_envelopes() method

This will be used in implementation of delivery status retention policy

wadcom committed:
Make remove_inactive_envelopes() return the number of affected rows

wadcom committed:
Add a cleaner process to implement retention policy for submissions

wadcom committed:
Mention sender address when simulating temporary failure

This is to aid in troubleshooting.

wadcom committed:
Loop through MXs when trying to deliver envelope

wadcom committed:
Add an end-to-end test for looping through MXs

wadcom committed:
Fix: missing "Content-Length" caused frontend to fail

wadcom committed:
Reject empty recipient list

wadcom committed:
When sending email to an MX, use only that domain's recipients

wadcom committed:
Remove obsolete TODO notes

wadcom committed:
Factor out _setup_proxy_methods_for()

horkhe commented on mailqueue.py:
```diff
@@ -0,0 +1,213 @@
+
+        for method in methods:
+            assert method not in self.__class__.__dict__, \
+                "name clash: method '{}' already exists in {}".format(method,
```
Not a question but a general comment. This seems like a nice trick that saves some typing... but in practice the code written using this trick is very hard to maintain. You can never obviously tell what methods a class has. IDEs do not understand that and therefore code completion and code insight do not work. Grep returns incomplete results... So even if we do not extend our offer please do not ever do that for your colleagues sake :)

wadcom replied:
I agree, this is not as straightforward as I would like. However, this is not to save typing.

The tradeoff here is this:

Option A: keep it simple, and manually keep parity between MailQueue and Manager's methods. Once you forget to mirror a method, this will only be discovered in runtime (Manager runs in its own thread, which makes it more difficult to test).

Option B: go with the more complex solution. Accept the need for higher mental effort to comprehend the code, forfeit IDE support.  Manager always mirrors MailQueue.

I started with option A, but as the number of methods in MailQueue grew, I found myself catching bugs of missing methods in end-to-end tests. At this stage (and in real life it would be CI/CD pipeline or staging, or production) bugs are much more expensive to fix, so I opted for more complex and less readable solution to gain in correctness and feedback time.

I think my colleagues would appreciate not being woken up during the night to find out I forgot to add a method to Manager. ;-)

So I agree with the general comment, but in this case I judged it was worth it (for me). If the team decides otherwise, the other option can always be implemented.

anton-efimenko replied:
In all honesty I've been there and I'd rather wake up, fix the missing method and go back to sleep. :-)

wadcom replied:
All right, I've reimplemented Manager with explicitly defined methods:

https://github.com/wadcom/mini-mailgun/pull/3/commits/6f08849eacf268f3aa4331a65632619521b42456

horkhe commented on sender.py:
```diff
@@ -0,0 +1,208 @@
+                                                                            delivered_to))
+            self._mailqueue.mark_as_sent(envelope)
+        else:
+            self._handle_temporary_failure(envelope)
```
What if some recipients from the envelope were rejected by the remote SMTP server, but others were accepted?

horkhe commented on sender.py:
```diff
@@ -0,0 +1,208 @@
+                                                                            delivered_to))
+            self._mailqueue.mark_as_sent(envelope)
+        else:
+            self._handle_temporary_failure(envelope)
```
What if the remote server returned 5xx (permanent failure) SMTP status?

wadcom replied:
Implemented overall logic to handle such conditions: 83b3e6d58af315010891bfde99def37dc8419aee.

Will handle per-recipient failures separately.

horkhe commented on cleaner.py:
```diff
@@ -0,0 +1,42 @@
+    setup_logging()
+
+    mq = mailqueue.MailQueue()
+    while True:
```
What is the reasoning behind having a cleaner process? An alternative would be to make Sender remove envelopes that are no more needed. When it completes a sending attempt it can make a decision whether an envelope should be retried or not.

wadcom replied:
My understanding of the requirements was that we want to keep the status for the retention period for *all* submissions, not only those that are still pending.

E.g. if I submit an email, it gets delivered, I would still be able to query its status (and get back "delivered") within 3 hours of my submission.

wadcom committed:
Update documentation and TODO

wadcom committed:
Explicitly define proxied methods in Manager

For simplicity and IDE support, per discussion here:

https://github.com/wadcom/mini-mailgun/pull/3#pullrequestreview-77016995

wadcom committed:
Add a special address to for the stub to permanently refuse sender

wadcom committed:
Handle permanent delivery failures

This covers broad SMTP errors (e.g. server refusing the sender
address), which are not recipient-specific. The logic to cover the
latter will be implemented separately.

wadcom committed:
Add an end-to-end test for permanent SMTP server failures

anton-efimenko commented:
Kudos for https://github.com/wadcom/mini-mailgun/pull/3/commits/6f08849eacf268f3aa4331a65632619521b42456
